### PR TITLE
[WIP] Invoice image

### PR DESCRIPTION
**Pull request by menyet**

### CODE REVIEW

File: server/ERNI.PBA.Server.Host/Model/PendingRequests/InvoiceImageModel.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+﻿using Microsoft.AspNetCore.Http;
+
+namespace ERNI.PBA.Server.Host.Model.PendingRequests
+{
+    public class InvoiceImageModel
+    {
+        public string RequestId { get; set; }
```
Comment (menyet) [inline server/ERNI.PBA.Server.Host/Model/PendingRequests/InvoiceImageModel.cs]: use int

File: server/ERNI.PBA.Server.Host/Model/PendingRequests/InvoiceImageModel.cs [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+        public IFormFile FileKey { get; set; }
```
Comment (menyet) [inline server/ERNI.PBA.Server.Host/Model/PendingRequests/InvoiceImageModel.cs]: why the name FileKey?

File: server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+        [HttpPost]
+        [Consumes("multipart/form-data")]
+        public async Task<IActionResult> AddInvoiceImage([FromForm] InvoiceImageModel invoiceImageModel,
+            CancellationToken cancellationToken)
+        {
+            byte[] buffer;
+
+            //Exclude from controller
+            using (var memStream = new MemoryStream())
```
Comment (menyet) [inline server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs]: what about

using (var openReadStream = formFile.OpenReadStream())
                {
                    var data = new byte[formFile.Length];
                    openReadStream.Read(data, 0, data.Length);

File: server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            {
+                await invoiceImageModel.FileKey.CopyToAsync(memStream, cancellationToken);
+                buffer = memStream.ToArray();
+            }
+
+            //Check if id is valid, and is in DB (415)
+            var requestId = Convert.ToInt32(invoiceImageModel.RequestId);
```
Comment (menyet) [inline server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs]: you do not need this, if you change RequestId to int

File: server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+            //Check if format is supported (415)
+            var fullName = invoiceImageModel.FileKey.FileName;
+            var extension = Path.GetExtension(fullName);
+            var fileName = Path.GetFileNameWithoutExtension(fullName);
```
Comment (menyet) [inline server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs]: not sure if it is necessary to separate filename and extension

File: server/ERNI.PBA.Server.DataAccess/Repository/InvoiceImageRepository.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    public class InvoiceImageRepository : IInvoiceImageRepository
+    {
+        private readonly DatabaseContext _context;
+
+        public InvoiceImageRepository(DatabaseContext context)
+        {
+            _context = context;
+        }
+
+        public async Task<string[]> GetInvoiceImagesName(int requestId, CancellationToken cancellationToken)
```
Comment (menyet) [inline server/ERNI.PBA.Server.DataAccess/Repository/InvoiceImageRepository.cs]: it is better to send also the ids of the images

File: server/ERNI.PBA.Server.Host/Model/PendingRequests/InvoiceImageModel.cs (hunk 1 above)
Comment (LubomirMaron) [reply server/ERNI.PBA.Server.Host/Model/PendingRequests/InvoiceImageModel.cs]: Wise

File: server/ERNI.PBA.Server.DataAccess/Repository/InvoiceImageRepository.cs [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    public class InvoiceImageRepository : IInvoiceImageRepository
+    {
+        private readonly DatabaseContext _context;
+
+        public InvoiceImageRepository(DatabaseContext context)
+        {
+            _context = context;
+        }
+
+        public async Task<Tuple<int,string>[]> GetInvoiceImagesNameId(int requestId, CancellationToken cancellationToken)
```
Comment (menyet) [inline server/ERNI.PBA.Server.DataAccess/Repository/InvoiceImageRepository.cs]: try to use names tuple here

File: server/ERNI.PBA.Server.Host/ERNI.PBA.Server.Host.csproj [hunk 8]
```diff
@@ -12,6 +12,7 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.AspNetCore.All" />
+    <PackageReference Include="Microsoft.AspNetCore.StaticFiles" Version="2.2.0" />
```
Comment (menyet) [inline server/ERNI.PBA.Server.Host/ERNI.PBA.Server.Host.csproj]: I think this reference isn't necessary

File: server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs (hunk 5 above)
Comment (LubomirMaron) [reply server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs]: Merged

File: server/ERNI.PBA.Server.Host/ERNI.PBA.Server.Host.csproj (hunk 8 above)
Comment (LubomirMaron) [reply server/ERNI.PBA.Server.Host/ERNI.PBA.Server.Host.csproj]: Used for TryGetContentType()

File: server/ERNI.PBA.Server.DataAccess/Repository/InvoiceImageRepository.cs (hunk 7 above)
Comment (LubomirMaron) [reply server/ERNI.PBA.Server.DataAccess/Repository/InvoiceImageRepository.cs]: Named tuples dont work well with EF Linq selec . Double conversion is needed

File: server/ERNI.PBA.Server.Host/Model/PendingRequests/InvoiceImageModel.cs [hunk 9]
```diff
@@ -0,0 +1,10 @@
+﻿using Microsoft.AspNetCore.Http;
+
+namespace ERNI.PBA.Server.Host.Model.PendingRequests
+{
+    public class InvoiceImageModel
```
Comment (menyet) [inline server/ERNI.PBA.Server.Host/Model/PendingRequests/InvoiceImageModel.cs]: rename to something like InvoiceImageModel

File: server/ERNI.PBA.Server.DataAccess/Repository/IInvoiceImageRepository.cs [hunk 10]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Threading;
+using System.Threading.Tasks;
+using ERNI.PBA.Server.DataAccess.Model;
+
+namespace ERNI.PBA.Server.DataAccess.Repository
+{
+    public interface IInvoiceImageRepository
+    {
+        Task<IdNamePair[]> GetInvoiceImagesIdNamePairs(int requestId, CancellationToken cancellationToken);
```
Comment (menyet) [inline server/ERNI.PBA.Server.DataAccess/Repository/IInvoiceImageRepository.cs]: Why IdNamePair? Call it InvoiceImageProjection, the method can be GetInvoiceImages

File: server/ERNI.PBA.Server.DataAccess/Repository/InvoiceImageRepository.cs [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+        }
+
+        public async Task AddInvoiceImage(InvoiceImage invoiceImage, CancellationToken cancellationToken)
+        {
+            await _context.InvoiceImages.AddAsync(invoiceImage, cancellationToken);
+        }
+
+        public async Task<InvoiceImage> GetInvoiceImage(int imageId, CancellationToken cancellationToken)
+        {
+            return await _context.InvoiceImages.FirstAsync(image => image.Id == imageId, cancellationToken);
```
Comment (menyet) [inline server/ERNI.PBA.Server.DataAccess/Repository/InvoiceImageRepository.cs]: Use firsOrDefault, then check for null in controller

File: server/ERNI.PBA.Server.DataAccess/Model/IdNamePair.cs [hunk 12]
```diff
@@ -0,0 +1,8 @@
+﻿namespace ERNI.PBA.Server.DataAccess.Model
```
Comment (menyet) [inline server/ERNI.PBA.Server.DataAccess/Model/IdNamePair.cs]: Let's move this into a separate namespace (projections?)

File: server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+                return StatusCode(413);
+            }
+
+            using (var openReadStream = invoiceImageModel.File.OpenReadStream())
+            {
+                buffer = new byte[invoiceImageModel.File.Length];
+                openReadStream.Read(buffer, 0, buffer.Length);
+            }
+
+            var image = new InvoiceImage
```
Comment (menyet) [inline server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs]: Input checks are missing

File: server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+            });
+            return Ok(result);
+        }
+
+        [HttpGet("image/{imageId}")]
+        [Authorize]
+        public async Task<IActionResult> GetInvoiceImageFile(int imageId,
+            CancellationToken cancellationToken)
+        {
+            var image = await _invoiceImageRepository.GetInvoiceImage(imageId, cancellationToken);
```
Comment (menyet) [inline server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs]: Input checks are missing

File: server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+        {
+            _invoiceImageRepository = invoiceImageRepository;
+            _unitOfWork = unitOfWork;
+        }
+
+        [HttpGet("images/{requestId}")]
+        [Authorize]
+        public async Task<IActionResult> GetInvoiceImages(int requestId, CancellationToken cancellationToken)
+        {
+            var imagesName = await _invoiceImageRepository.GetInvoiceImagesIdNamePairs(requestId, cancellationToken);
```
Comment (menyet) [inline server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs]: Input checks are missing

File: client/src/app/file-upload/file-upload.component.ts [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+      let newItem = new uploadingImage();
+
+      newItem.name = image.file.name;
+      newItem.progress = 0;
+      this.uploadingImages.push(newItem);
+
+      this.invoiceImageService.addInvoiceImage(image).subscribe(progress => {
+        newItem.progress = progress;
+      }, (error) => {
+        this.fileIsTooLarge = true;
```
Comment (menyet) [inline client/src/app/file-upload/file-upload.component.ts]: Is 413 the onlz possible error?

File: client/src/app/file-upload/file-upload.component.ts [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+  requestId: number;
+  fileIsTooLarge : boolean = false;
+
+  constructor(private invoiceImageService: InvoiceImageService) { }
+
+  ngOnInit() {  }
+
+  ngOnChanges(changes: SimpleChanges) {
+
+    if (!changes.requestIdInput || changes.requestIdInput.currentValue == undefined) {
```
Comment (menyet) [inline client/src/app/file-upload/file-upload.component.ts]: Do this using observables in ngOnInit

Comment (LubomirMaron) [reply client/src/app/file-upload/file-upload.component.ts]: During ngOnInit requestIdInput is not yet initialized.

File: client/src/app/file-upload/file-upload.component.ts (hunk 16 above)
Comment (LubomirMaron) [reply client/src/app/file-upload/file-upload.component.ts]: Changed to generic error

File: server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs (hunk 15 above)
Comment (LubomirMaron) [reply server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs]: How to check int ?

File: server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+        }
+
+        [HttpGet("image/{imageId}")]
+        [Authorize]
+        public async Task<IActionResult> GetInvoiceImageFile(int imageId,
+            CancellationToken cancellationToken)
+        {
+            var image = await _invoiceImageRepository.GetInvoiceImage(imageId, cancellationToken);
+            var userId = HttpContext.User.GetId();
+            if (!await _permissionHelper.IsUsersRequestOrAdmin(userId, image.RequestId, cancellationToken))
```
Comment (menyet) [inline server/ERNI.PBA.Server.Host/Controllers/InvoiceImageController.cs]: You do not need a helper class there. Check deleteRequest method:
var user = HttpContext.User;
            if (!user.IsInRole(Roles.Admin) && user.GetId() != request.UserId)
            {
                return BadRequest("Access denied");
            }

**Pull request merged**